debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlValueConverters.java

```diff
@@ -308,32 +323,50 @@ protected boolean matches(String upperCaseTypeName, String upperCaseMatch) {
     }
 
-    protected String extractEnumAndSetOptions(Column column, boolean commaSeparated) {
+    protected String extractEnumAndSetOptions(Column column) {
         String options = MySqlDdlParser.parseSetAndEnumOptions(column.typeExpression());
-        if (!commaSeparated) return options;
-        StringBuilder sb = new StringBuilder();
-        boolean first = true;
-        for (int i = 0; i != options.length(); ++i) {
-            if (first)
-                first = false;
-            else
-                sb.append(',');
-            sb.append(options.charAt(i));
-        }
-        return sb.toString();
+        return options;
     }
 
     protected String convertSetValue(long indexes, String options) {
         StringBuilder sb = new StringBuilder();
         int index = 0;
         boolean first = true;
+        HashMap<Integer, Integer> deliminatorMap = new HashMap<Integer, Integer>();
         while (indexes != 0L) {
             if (indexes % 2L != 0) {
                 if (first)
                     first = false;
                 else
                     sb.append(',');
-                sb.append(options.substring(index, index + 1));
+                int setIndex = index;
+                /*Playing with index is better, than splitting it into list array each time.
+                  Though it means handle error/edge cases
+                  Can be optimized using hash map as we are calculating index again and again */
+                int startDeliminatorIndex = 0;
+                int nextDelimiatorIndex = 0;
+                if (deliminatorMap.containsKey(startDeliminatorIndex)) {
+                    nextDelimiatorIndex = deliminatorMap.get(startDeliminatorIndex);
+                }else {
+                    nextDelimiatorIndex = options.indexOf(MySqlDdlParser.ENUM_AND_SET_DELIMINATOR);
+                    deliminatorMap.put(startDeliminatorIndex,nextDelimiatorIndex);
+                }
+                while (setIndex > 0 && nextDelimiatorIndex != -1) {
+                    startDeliminatorIndex = nextDelimiatorIndex + 1;
+                    if (deliminatorMap.containsKey(startDeliminatorIndex)) {
+                        nextDelimiatorIndex = deliminatorMap.get(startDeliminatorIndex);
+                    }else {
+                        nextDelimiatorIndex = options.indexOf(MySqlDdlParser.ENUM_AND_SET_DELIMINATOR,startDeliminatorIndex);
+                        deliminatorMap.put(startDeliminatorIndex,nextDelimiatorIndex);
+                    }
+                    setIndex --;
+                }
+                //in case of last enum from the comma seperated string
+                if (setIndex == 0 && nextDelimiatorIndex == -1) {
+                    sb.append(options.substring(startDeliminatorIndex));
+                } else if (nextDelimiatorIndex != -1) {
+                    sb.append(options.substring(startDeliminatorIndex,nextDelimiatorIndex));
+                }
```

Comment:
This is a method that is called with *every row change*, so if the options were passed in as a `List<String>` and this method becomes far simpler, it will also perform much better.